storage/olddbclean/oldDatabaseCleaner.go

```diff
@@ -0,0 +1,231 @@
+}
+
+type oldDatabaseCleaner struct {
+	databasePath        string
+	storageListProvider StorageListProviderHandler
+	fileRemover         func(file string) error
+	directoryReader     storage.DirectoryReaderHandler
+	oldestEpochsToKeep  map[uint32]uint32
+
+	sync.RWMutex
+}
+
+// NewOldDatabaseCleaner returns a new instance of oldDatabaseCleaner
+func NewOldDatabaseCleaner(args Args) (*oldDatabaseCleaner, error) {
+	if check.IfNil(args.StorageListProvider) {
+		return nil, storage.ErrNilStorageListProvider
+	}
+	if check.IfNil(args.EpochStartNotifier) {
+		return nil, storage.ErrNilEpochStartNotifier
+	}
+
+	fileRemoverFunc := func(file string) error {
+		return os.RemoveAll(file)
+	}
+	directoryReader := directoryhandler.NewDirectoryReader()
+
+	odc := &oldDatabaseCleaner{
+		databasePath:        args.DatabasePath,
+		storageListProvider: args.StorageListProvider,
+		fileRemover:         fileRemoverFunc,
+		directoryReader:     directoryReader,
+		oldestEpochsToKeep:  make(map[uint32]uint32),
+	}
+
+	odc.registerHandler(args.EpochStartNotifier)
+
+	return odc, nil
+}
+
+// registerHandler will register a new function to the epoch start notifier
+func (odc *oldDatabaseCleaner) registerHandler(handler EpochStartNotifier) {
+	subscribeHandler := notifier.NewHandlerForEpochStart(
+		func(hdr data.HeaderHandler) {
+			err := odc.handleEpochChangePrepare(hdr.GetEpoch())
+			if err != nil {
+				log.Warn("oldDatabaseCleaner: handleEpochChangePrepare", "error", err)
+			}
+		},
+		func(hdr data.HeaderHandler) {},
+		core.OldDatabaseCleanOrder)
+
+	handler.RegisterHandler(subscribeHandler)
+}
+
+func (odc *oldDatabaseCleaner) handleEpochChangePrepare(epoch uint32) error {
+	newOldestEpoch := odc.computeOldestEpochToKeep(epoch)
+	odc.oldestEpochsToKeep[epoch] = newOldestEpoch
+	log.Debug("old database cleaner", "epoch", epoch, "inner map", odc.oldestEpochsToKeep)
+	shouldClean := odc.shouldCleanOldData(epoch, newOldestEpoch)
+	if !shouldClean {
+		return nil
+	}
+
+	err := odc.cleanOldEpochs(epoch)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (odc *oldDatabaseCleaner) shouldCleanOldData(currentEpoch uint32, newOldestEpoch uint32) bool {
+	odc.RLock()
+	defer odc.RUnlock()
+
+	epochForDeletion := currentEpoch - 1
+	previousOldestEpoch, epochExists := odc.oldestEpochsToKeep[epochForDeletion]
+	if !epochExists {
+		log.Debug("cannot delete old databases as the previous epoch does not exist in configuration",
+			"epoch", epochForDeletion,
+			"epochs configuration", odc.oldestEpochsToKeep)
+		return false
+	}
+	if previousOldestEpoch > newOldestEpoch {
+		log.Debug("skipping cleaning of old databases because new oldest epoch is older than previous",
+			"previous older epoch", previousOldestEpoch,
+			"actual older epoch", newOldestEpoch)
+		return false
+	}
+
+	return true
+}
+
+func (odc *oldDatabaseCleaner) computeOldestEpochToKeep(currentEpoch uint32) uint32 {
+	odc.Lock()
+	defer odc.Unlock()
+
+	oldestEpoch := uint32(math.MaxUint32)
```

Comment:
why not write this:
```GO
oldestEpoch = currentEpoch
```

and remove the if branch on L138 ?

Reply:
refactored as the old version wasn't correct. It considered the newest (current) epoch to be the oldest in case no storer sends a valid oldest epoch. Now the function returns error if any situation like this